### PR TITLE
[luci/pass] Add more unit tests for ResovleFormerCustomPass

### PR DESCRIPTION
**Pull request by Seunghui98**
This commit adds more unit tests for ResolveFormerCustomPass.

ONE-DCO-1.0-Signed-off-by: SeungHui Lee <shsh1004.lee@samsung.com>

---
Related to Issue: https://github.com/Samsung/ONE/issues/10784
Draft PR: [#10737](https://github.com/Samsung/ONE/pull/11701)

### CODE REVIEW

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 1]
```diff
@@ -16,11 +16,136 @@
 
 #include "luci/Pass/ResolveFormerCustomOpPass.h"
```
Comment (Seunghui98) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:18]: The related content of this PR is here. (https://github.com/Samsung/ONE/pull/12318#discussion_r1436660933)

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 2]
```diff
@@ -16,11 +16,136 @@
...
+class FormerCustomOpGraphTest : public ::testing::Test
 {
+public:
+  BroadcastToGraph g;
   luci::ResolveFormerCustomOpPass pass;
+};
+
+} // namespace
+
+TEST_F(FormerCustomOpGraphTest, name)
```
Comment (seanshpark) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:114]: plz split TEST to TEST_F change to another PR.

**Comment by seanshpark** [comment]
Let's revise title and commit message to
```
[luci/pass] Add more unit tests for ResovleFormerCustomPass

This commit adds more unit tests for ResolveFormerCustomPass
```

**Comment by seanshpark** [comment]
@Seunghui98 , can you distinguish between "Positive" and "Negative" tests?

**Comment by Seunghui98** [comment]
@seanshpark 

> can you distinguish between "Positive" and "Negative" tests?

yes. :)
As far as I know, "Positive" test is a test in which normal expected results come out with valid inputs, and "Negative" test is a test in which abnormal results come out with invalid inputs.

**Comment by seanshpark** [comment]
> "Negative" test is a test in which abnormal results come out with invalid inputs.

OK, good :)
then, can you add some more negative tests for invalid inputs?
we don't have to introduce for all invalid cases but one or two more would be nice.

**Comment by Seunghui98** [comment]
@seanshpark 

> then, can you add some more negative tests for invalid inputs?
yes. I will add invalid cases.
Thank you a lot :)

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 3]
```diff
@@ -16,11 +16,162 @@
...
+{
+  g.init();
+
+  auto ret = pass.run(g.g());
+  EXPECT_EQ(true, ret);
+
+  auto broadcastTo = dynamic_cast<luci::CircleBroadcastTo *>(g.output(0)->from());
+  EXPECT_NE(nullptr, broadcastTo);
+
+  auto input = loco::must_cast<luci::CircleConst *>(broadcastTo->input());
```
Comment (seanshpark) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:135]: would be better to use dynamic cast

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 4]
```diff
@@ -16,11 +16,162 @@
...
+  EXPECT_NE(nullptr, input);
+  EXPECT_EQ(1, input->at<loco::DataType::FLOAT32>(0));
+  EXPECT_EQ(2, input->at<loco::DataType::FLOAT32>(1));
+  EXPECT_EQ(3, input->at<loco::DataType::FLOAT32>(2));
+
+  auto shape = loco::must_cast<luci::CircleConst *>(broadcastTo->shape());
```
Comment (seanshpark) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:141]: ditto

**Comment by Seunghui98** [comment]
@seanshpark @mhs4670go 

I reflected comments at this PR.
Could you review it again..?

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 5]
```diff
@@ -16,11 +16,162 @@
...
+ *
+ *     [Const(Input)] [Const(Shape)]
+ *             \         /
+ *         [Custom(BroadcastTo)]
+ *                  |
+ *             [CustomOut]
+ *                  |
+ *              [Output]
+ */
+class BroadcastToGraphlet
```
Comment (seanshpark) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:40]:
```suggestion
class CustomBroadcastToGraphlet
```

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 6]
```diff
@@ -16,11 +16,162 @@
...
+
+    output(0)->from(_broadcastTo_out);
+  }
+};
+
+class FormerCustomOpGraphTest : public ::testing::Test
+{
+public:
+  BroadcastToGraph g;
+  luci::ResolveFormerCustomOpPass pass;
```
Comment (seanshpark) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:113]:
```suggestion
  BroadcastToGraph _g;
  luci::ResolveFormerCustomOpPass _pass;
```

**Review by seanshpark — DISMISSED** [review]
LGTM thank you!

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 7]
```diff
@@ -16,11 +16,164 @@
...
+
+  auto input = dynamic_cast<luci::CircleConst *>(broadcastTo->input());
+  EXPECT_NE(nullptr, input);
+  EXPECT_EQ(1, input->at<loco::DataType::FLOAT32>(0));
+  EXPECT_EQ(2, input->at<loco::DataType::FLOAT32>(1));
+  EXPECT_EQ(3, input->at<loco::DataType::FLOAT32>(2));
+
+  auto shape = dynamic_cast<luci::CircleConst *>(broadcastTo->shape());
+  EXPECT_NE(nullptr, shape);
+  EXPECT_EQ(true, (shape->dtype() == loco::DataType::S32 || shape->dtype() == loco::DataType::S64));
```
Comment (mhs4670go) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:145]: Maybe it should set exact type.
```suggestion
  EXPECT_EQ(true, (shape->dtype() == loco::DataType::S32));
```

File: compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp [hunk 8]
```diff
@@ -16,11 +16,164 @@
...
+  EXPECT_EQ(2, shape->at<loco::DataType::S32>(0));
+  EXPECT_EQ(3, shape->at<loco::DataType::S32>(1));
+}
+
+TEST_F(FormerCustomOpGraphTest, wrong_op_NEG)
+{
+  _g.init();
+
+  _g.broadcastTo()->custom_code("Abs");
```
Comment (mhs4670go) [inline compiler/luci/pass/src/ResolveFormerCustomOpPass.test.cpp:154]: I think the test should be like below because this test would be broken if `Abs` custom code become supported.
```suggestion
  _g.broadcastTo()->custom_code("UNSUPORTED_CUSTOM_CODE");
```

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
